# **Computing BLEU SCORE**

In [ ]:
from nltk.translate.bleu_score import corpus_bleu

In [ ]:
# Compute BLEU score
def compute_BLEU_score(output, ground_truth):
    bleu_score = corpus_bleu([[translation] for translation in ground_truth], output)
    print("BLEU Score:", bleu_score)


# **ComputingROUGE SCORE**

In [ ]:
!pip install rouge

In [ ]:
from rouge import Rouge

In [ ]:
# Initialize Rouge
rouge = Rouge()

In [ ]:
# Calculate ROUGE scores
def compute_ROUGE_score(ground_truth,output):
    scores = rouge.get_scores(output, ground_truth, avg=True)
    print("ROUGE Scores:", scores)


# **1.1) NLLB-200 with 600M parameters**

In [ ]:
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer
import os
import torch

In [ ]:
# Assuming your dataset is under /kaggle/input/dataset_language folder
# dataset_folder = "/kaggle/input/language-dataset"

# Load tokenizer and model
model = AutoModelForSeq2SeqLM.from_pretrained("facebook/nllb-200-distilled-600M")
# Read English sentences from test_en file
test_en_file = os.path.join("test.en")
english_sentences=[]
gujarati_sentences=[]
marathi_sentences=[]
with open(test_en_file, "r", encoding="utf-8") as file:
    english_sentences = [line.strip() for line in file.readlines()]
#     english_sentences = file.readlines()

test_gu_file = os.path.join("test.gu")
with open(test_gu_file, "r", encoding="utf-8") as file:
    gujarati_sentences = [line.strip() for line in file.readlines()]

test_mr_file = os.path.join("test.mr")
with open(test_mr_file, "r", encoding="utf-8") as file:
    marathi_sentences = [line.strip() for line in file.readlines()]

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/846 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/2.46G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/189 [00:00<?, ?B/s]

In [ ]:
len(english_sentences)

2390

In [ ]:
len(gujarati_sentences)

2390

In [ ]:
len(marathi_sentences)

2390

In [ ]:

import random
# Generate 1000 random numbers within the range of 1 to 2390
# random_numbers = [random.randint(1, 2390) for _ in range(1000)]
random_numbers = [2036,
 1352,
 961,
 334,
 1178,
 1260,
 1455,
 552,
 1616,
 698,
 2089,
 1333,
 2041,
 1393,
 1985,
 1321,
 1269,
 50,
 1300,
 210,
 448,
 386,
 629,
 1661,
 873,
 1226,
 810,
 775,
 451,
 2244,
 2276,
 1393,
 1689,
 1545,
 659,
 1273,
 950,
 1058,
 330,
 215,
 1290,
 1011,
 676,
 811,
 2056,
 1213,
 1093,
 333,
 2298,
 1556,
 66,
 475,
 1395,
 1251,
 976,
 1108,
 64,
 2067,
 81,
 2028,
 1583,
 1272,
 1483,
 650,
 431,
 19,
 2329,
 842,
 438,
 1062,
 69,
 560,
 110,
 1001,
 2212,
 165,
 772,
 1520,
 2175,
 837,
 51,
 253,
 1567,
 1032,
 834,
 1288,
 671,
 1988,
 190,
 140,
 296,
 2049,
 2106,
 1535,
 1823,
 1521,
 897,
 1488,
 931,
 1745,
 1667,
 29,
 1111,
 589,
 2086,
 667,
 1073,
 1694,
 1888,
 1360,
 1533,
 1193,
 951,
 1870,
 834,
 1308,
 34,
 18,
 800,
 1020,
 142,
 1142,
 1849,
 1567,
 2336,
 202,
 1168,
 1192,
 725,
 326,
 234,
 579,
 2293,
 614,
 1508,
 311,
 1630,
 1891,
 1030,
 1304,
 135,
 364,
 234,
 255,
 1443,
 1481,
 145,
 491,
 1508,
 2368,
 668,
 1692,
 881,
 1916,
 20,
 1140,
 263,
 1625,
 1377,
 1892,
 668,
 1183,
 792,
 1394,
 1112,
 2169,
 1541,
 754,
 499,
 2201,
 1210,
 1572,
 2232,
 461,
 906,
 385,
 1730,
 782,
 1976,
 840,
 831,
 1774,
 812,
 883,
 175,
 1407,
 1900,
 925,
 1697,
 614,
 397,
 617,
 219,
 2101,
 1926,
 2090,
 1762,
 646,
 1592,
 1625,
 997,
 2264,
 1668,
 733,
 567,
 657,
 2033,
 769,
 1418,
 2313,
 2360,
 1370,
 2101,
 1791,
 1470,
 617,
 418,
 1282,
 1596,
 1248,
 978,
 1158,
 1796,
 2031,
 1667,
 1324,
 1998,
 174,
 1564,
 2361,
 337,
 1621,
 1256,
 1690,
 2285,
 1986,
 1284,
 815,
 2028,
 743,
 1570,
 1685,
 1212,
 266,
 835,
 712,
 238,
 689,
 2202,
 670,
 2103,
 1225,
 2225,
 131,
 131,
 2325,
 1297,
 519,
 997,
 1733,
 1146,
 1792,
 2197,
 1761,
 1351,
 591,
 1648,
 2188,
 1363,
 461,
 764,
 271,
 1252,
 1662,
 1333,
 1336,
 1595,
 1451,
 616,
 1300,
 224,
 1822,
 1267,
 47,
 1526,
 793,
 1104,
 195,
 2123,
 936,
 977,
 481,
 638,
 853,
 320,
 1347,
 494,
 1456,
 803,
 294,
 447,
 1922,
 1957,
 497,
 2136,
 404,
 1729,
 1128,
 933,
 495,
 1877,
 1755,
 1293,
 1542,
 13,
 138,
 2241,
 2340,
 645,
 1225,
 475,
 951,
 2270,
 427,
 1847,
 2059,
 2044,
 2155,
 1415,
 1277,
 1384,
 1708,
 821,
 2131,
 211,
 2202,
 1026,
 1745,
 1361,
 2138,
 1711,
 1983,
 202,
 1156,
 1263,
 680,
 416,
 585,
 980,
 2384,
 707,
 756,
 1680,
 1330,
 662,
 1290,
 355,
 757,
 2367,
 786,
 1291,
 1532,
 1474,
 1173,
 974,
 1514,
 529,
 141,
 72,
 1630,
 497,
 549,
 1341,
 1830,
 686,
 298,
 124,
 958,
 1174,
 1716,
 967,
 1058,
 932,
 423,
 1958,
 2054,
 1383,
 201,
 2280,
 2281,
 1056,
 317,
 1296,
 1146,
 2004,
 1483,
 1890,
 364,
 300,
 732,
 1503,
 270,
 1577,
 2038,
 1877,
 597,
 63,
 1459,
 1710,
 99,
 1298,
 1174,
 989,
 1489,
 164,
 1190,
 1868,
 412,
 585,
 808,
 570,
 730,
 972,
 570,
 327,
 1767,
 734,
 1213,
 1382,
 800,
 189,
 188,
 1943,
 1735,
 646,
 1047,
 2266,
 1114,
 1369,
 569,
 1258,
 609,
 1626,
 1760,
 2163,
 1668,
 1054,
 337,
 439,
 224,
 348,
 1101,
 903,
 1340,
 2336,
 2175,
 2375,
 2210,
 1000,
 1600,
 613,
 233,
 1813,
 1049,
 542,
 1706,
 2385,
 1992,
 438,
 206,
 462,
 1916,
 2301,
 478,
 1086,
 2153,
 795,
 450,
 1483,
 844,
 571,
 2278,
 1945,
 1927,
 1650,
 2189,
 1978,
 19,
 419,
 550,
 76,
 944,
 1123,
 367,
 1902,
 769,
 1933,
 484,
 1529,
 1180,
 1823,
 865,
 630,
 1172,
 699,
 1902,
 518,
 433,
 1846,
 686,
 1782,
 1601,
 165,
 981,
 258,
 1276,
 2180,
 760,
 2375,
 1900,
 2299,
 1095,
 1500,
 2026,
 678,
 1619,
 205,
 926,
 943,
 2387,
 1872,
 1047,
 1438,
 1839,
 929,
 1759,
 1240,
 1606,
 2209,
 1353,
 1837,
 692,
 2151,
 843,
 1340,
 68,
 580,
 1034,
 592,
 2236,
 1752,
 1244,
 2266,
 1813,
 224,
 1109,
 2119,
 275,
 680,
 779,
 2357,
 1957,
 2047,
 601,
 382,
 2068,
 2007,
 1466,
 1897,
 807,
 2271,
 2376,
 162,
 967,
 924,
 1388,
 1131,
 1215,
 665,
 1508,
 603,
 306,
 1291,
 1157,
 311,
 171,
 1503,
 400,
 1466,
 845,
 2050,
 1382,
 601,
 834,
 1066,
 1025,
 1396,
 992,
 67,
 1067,
 1512,
 346,
 337,
 28,
 1021,
 781,
 1941,
 1756,
 1672,
 2043,
 313,
 462,
 533,
 1482,
 1827,
 1681,
 1117,
 745,
 54,
 2179,
 663,
 2257,
 892,
 851,
 2087,
 2332,
 1326,
 701,
 1655,
 17,
 1645,
 107,
 1691,
 630,
 746,
 1671,
 479,
 1921,
 334,
 364,
 1197,
 1299,
 872,
 212,
 1259,
 2344,
 1894,
 1998,
 2276,
 1345,
 695,
 506,
 1450,
 375,
 1531,
 2036,
 1467,
 538,
 1686,
 1232,
 1232,
 245,
 1725,
 1265,
 1421,
 1888,
 941,
 1663,
 2133,
 1022,
 2104,
 2226,
 790,
 1108,
 1399,
 1853,
 1716,
 1564,
 243,
 1426,
 1853,
 2083,
 394,
 2056,
 786,
 134,
 1834,
 2066,
 911,
 2181,
 1986,
 39,
 1579,
 2,
 924,
 1476,
 1748,
 53,
 299,
 1135,
 2294,
 484,
 857,
 2123,
 2000,
 1042,
 780,
 1010,
 130,
 1534,
 1296,
 468,
 1746,
 465,
 1763,
 926,
 444,
 1005,
 2089,
 251,
 2125,
 2299,
 2018,
 30,
 879,
 713,
 1426,
 1935,
 1749,
 1592,
 1207,
 512,
 2207,
 1517,
 2212,
 1238,
 421,
 1172,
 49,
 1936,
 1723,
 629,
 1427,
 2117,
 2142,
 465,
 2197,
 1887,
 2189,
 274,
 1747,
 517,
 1019,
 9,
 82,
 138,
 1079,
 1756,
 2009,
 66,
 1525,
 837,
 1728,
 1927,
 383,
 2007,
 1171,
 602,
 187,
 2114,
 1014,
 265,
 751,
 1569,
 570,
 932,
 414,
 91,
 1379,
 1907,
 1737,
 720,
 2189,
 1056,
 1545,
 1390,
 1625,
 1310,
 1793,
 794,
 1161,
 2390,
 1606,
 1706,
 1150,
 1479,
 2329,
 685,
 996,
 2296,
 460,
 698,
 957,
 731,
 390,
 295,
 2057,
 575,
 1991,
 1282,
 1419,
 1970,
 2174,
 1517,
 15,
 1513,
 632,
 578,
 1946,
 527,
 341,
 2386,
 278,
 1835,
 2267,
 561,
 1241,
 1379,
 1089,
 1476,
 183,
 340,
 2223,
 1365,
 128,
 27,
 580,
 930,
 1719,
 1833,
 1564,
 2360,
 504,
 1840,
 1533,
 654,
 1944,
 236,
 123,
 2335,
 2208,
 2134,
 2241,
 2315,
 1075,
 1335,
 1450,
 816,
 195,
 948,
 991,
 1187,
 1484,
 234,
 1549,
 967,
 697,
 160,
 497,
 1733,
 523,
 96,
 173,
 1104,
 2144,
 2108,
 1652,
 1610,
 542,
 1181,
 2248,
 1171,
 2177,
 1051,
 2126,
 2321,
 347,
 227,
 1632,
 1525,
 1840,
 2068,
 1143,
 2061,
 1683,
 1730,
 1693,
 1509,
 1588,
 1338,
 290,
 660,
 1388,
 1310,
 1844,
 583,
 1205,
 804,
 806,
 2307,
 627,
 1948,
 1126,
 1628,
 296,
 1603,
 1370,
 323,
 1515,
 273,
 1034,
 55,
 391,
 1864,
 2350,
 255,
 1462,
 1748,
 900,
 2214,
 387,
 1085,
 292,
 1057,
 329,
 999,
 653,
 284,
 1487,
 224,
 1084,
 1600,
 33,
 2074,
 2063,
 1635,
 2043,
 2169,
 847,
 1492,
 2327,
 18,
 1672,
 1629,
 1148,
 993,
 508,
 1726,
 2342,
 2190,
 1894,
 474,
 1484,
 25,
 2098,
 634,
 1209,
 1842,
 637,
 2150,
 2390,
 1545,
 1360,
 908,
 54,
 837,
 1807,
 1207,
 493,
 406,
 273,
 2221,
 2151,
 1098,
 1441,
 249,
 1785,
 1440,
 276,
 1575,
 1619,
 631,
 2014,
 534,
 522,
 2197,
 1022,
 2027,
 1169,
 1928]


# Select 1000 random sentences from the list of English sentences
english_sentences = [english_sentences[index - 1] for index in random_numbers]
gujarati_sentences = [gujarati_sentences[index - 1] for index in random_numbers]
marathi_sentences = [marathi_sentences[index - 1] for index in random_numbers]

In [ ]:
print(len(english_sentences))
# print(gujarati_sentences)
# print(marathi_sentences)

1000


In [ ]:
def translate_from_source_to_target(source_lang,target_lang,source_sentences):
    # Translate Gujarati sentences to English
    output_sentences = []
    tokenizer = AutoTokenizer.from_pretrained(
        "facebook/nllb-200-distilled-600M", src_lang=source_lang
    )
    i=0
    for sentence in source_sentences:
        # print(sentence)
        print(i)
        i=i+1
        inputs = tokenizer(sentence, return_tensors="pt")

        translated_tokens = model.generate(
            **inputs, forced_bos_token_id=tokenizer.lang_code_to_id[target_lang], max_length=30
        )
        translated_sentence=tokenizer.batch_decode(translated_tokens, skip_special_tokens=True)[0]
    #     print(translated_sentence)
        output_sentences.append(translated_sentence)
    return output_sentences

#  1) English to Gujarati

In [ ]:
# Set the source and target languages
source_lang = "eng_Latn"
target_lang = "guj_Gujr"
eng_to_gu = translate_from_source_to_target(source_lang,target_lang,english_sentences)

tokenizer_config.json:   0%|          | 0.00/564 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/4.85M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.3M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/3.55k [00:00<?, ?B/s]

the `lang_code_to_id` attribute is deprecated. The logic is natively handled in the `tokenizer.adder_tokens_decoder` this attribute will be removed in `transformers` v4.38


0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

In [ ]:
eng_to_gu[:5]

['તેઓ પાસે અનન્ય શક્તિઓ અને સમજશક્તિઓ પણ છે.',
 'પ્રધાનમંત્રીએ રેલવે, રોડ, પેટ્રોલિયમ અને વીજ ક્ષેત્રોમાં નવ માળખાગત પ્રોજેક્ટ્સની પ્રગતિ',
 'ગયા મહિને, તેઓએ મારી મન કી બાતના પ્રસારણ પછી તરત જ પ્રાદેશિક ભાષાઓમાં સંસ્કરણો પ્રસાર',
 'ગયા વર્ષે અમે એસસીઓનાં તાશ્કંદ બેઠકમાં પૂર્ણ સભ્યપદની પ્રક્રિયા શરૂ કરી હતી.',
 'આ સમગ્ર રાષ્ટ્ર આ સમય દરમિયાન પૂર્વોત્તરના લોકો સાથે છે.']

In [ ]:
# Open a new text file for writing
with open('eng_to_gu.txt', 'w') as file:
    # Iterate over each string in the list
    for string in eng_to_gu:
        # Write the string to the file followed by a newline character
        file.write(string + '\n')


In [ ]:
# # Initialize an empty list to store the sentences
# eng_to_gu = []

# # Open the text file for reading
# with open('/content/eng_to_gu.txt', 'r') as file:
#     # Read each line (sentence) from the file
#     for line in file:
#         # Append the line (sentence) to the list, removing any leading/trailing whitespace
#         eng_to_gu.append(line.strip())

# # Print the list of sentences to verify
# print(eng_to_gu[:5])


['મને એ નોંધતા આનંદ થાય છે કે ભારત અને બ્રાઝિલએ દ્વિપક્ષીય રોકાણ કરારનો લખાણ અંતિમ સ્વ', 'તાજેતરમાં 30 નાનો ઉપગ્રહો લોન્ચ કરવામાં આવ્યા હતા.', 'પ્રધાનમંત્રીએ જણાવ્યું હતું કે, સ્વચ્છ ભારત મિશન, ડિજિટલ વ્યવહારો અને કૌશલ્ય વિકાસ જેવા મુદ્દાઓ પર', 'પ્રધાનમંત્રી  ટેકનોલોજી અને ગ્રામીણ જીવન પર પ્રદર્શનની મુલાકાત લેશે.', 'પ્રધાનમંત્રી  ટેકનોલોજી અને ગ્રામીણ જીવન પર પ્રદર્શનની મુલાકાત લેશે.']


# **2) Gujarati To English**

In [ ]:
random_numbers

[2036,
 1352,
 961,
 334,
 1178,
 1260,
 1455,
 552,
 1616,
 698,
 2089,
 1333,
 2041,
 1393,
 1985,
 1321,
 1269,
 50,
 1300,
 210,
 448,
 386,
 629,
 1661,
 873,
 1226,
 810,
 775,
 451,
 2244,
 2276,
 1393,
 1689,
 1545,
 659,
 1273,
 950,
 1058,
 330,
 215,
 1290,
 1011,
 676,
 811,
 2056,
 1213,
 1093,
 333,
 2298,
 1556,
 66,
 475,
 1395,
 1251,
 976,
 1108,
 64,
 2067,
 81,
 2028,
 1583,
 1272,
 1483,
 650,
 431,
 19,
 2329,
 842,
 438,
 1062,
 69,
 560,
 110,
 1001,
 2212,
 165,
 772,
 1520,
 2175,
 837,
 51,
 253,
 1567,
 1032,
 834,
 1288,
 671,
 1988,
 190,
 140,
 296,
 2049,
 2106,
 1535,
 1823,
 1521,
 897,
 1488,
 931,
 1745,
 1667,
 29,
 1111,
 589,
 2086,
 667,
 1073,
 1694,
 1888,
 1360,
 1533,
 1193,
 951,
 1870,
 834,
 1308,
 34,
 18,
 800,
 1020,
 142,
 1142,
 1849,
 1567,
 2336,
 202,
 1168,
 1192,
 725,
 326,
 234,
 579,
 2293,
 614,
 1508,
 311,
 1630,
 1891,
 1030,
 1304,
 135,
 364,
 234,
 255,
 1443,
 1481,
 145,
 491,
 1508,
 2368,
 668,
 1692,
 881,
 1916,
 

In [ ]:
# Set the source and target languages
target_lang = "eng_Latn"
source_lang = "guj_Gujr"
gu_to_en = translate_from_source_to_target(source_lang,target_lang,gujarati_sentences)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

In [ ]:
gu_to_en[:5]

['They also have special abilities and insight.',
 'The Prime Minister reviewed the progress made in the nine-phase construction projects in the railway, road, petroleum and energy',
 'Since last month, he has been giving Mention in regional languages as soon as I have finished speaking.',
 'We started the process of giving full membership to India at the SCO meeting held in Tashkent last year.',
 'The whole country is with the people of the North East in this difficult time.']

In [ ]:
# Open a new text file for writing
with open('gu_to_en.txt', 'w') as file:
    # Iterate over each string in the list
    for string in gu_to_en:
        # Write the string to the file followed by a newline character
        file.write(string + '\n')

# **3) Gujarati To Marathi**

In [ ]:
# Set the source and target languages
target_lang = "mar_Deva"
source_lang = "guj_Gujr"
gu_to_mr = translate_from_source_to_target(source_lang,target_lang,gujarati_sentences)

In [ ]:
gu_to_mr[:5]

In [ ]:
# Open a new text file for writing
with open('gu_to_mr.txt', 'w') as file:
    # Iterate over each string in the list
    for string in gu_to_mr:
        # Write the string to the file followed by a newline character
        file.write(string + '\n')

# **3) Marathi To Gujarati**

In [ ]:
# Set the source and target languages
target_lang = "mar_Deva"
source_lang = "guj_Gujr"
mr_to_gu = translate_from_source_to_target(source_lang,target_lang,marathi_sentences)

In [ ]:
mr_to_gu[:5]

In [ ]:
# Open a new text file for writing
with open('mr_to_gu.txt', 'w') as file:
    # Iterate over each string in the list
    for string in mr_to_gu:
        # Write the string to the file followed by a newline character
        file.write(string + '\n')

# **1.2) IndicTrans2**

## Setup

Please run the cells below to install the necessary dependencies.


In [ ]:
%%capture
!git clone https://github.com/AI4Bharat/IndicTrans2.git

In [ ]:
%%capture
%cd /content/IndicTrans2/huggingface_interface

In [ ]:
%%capture
!python3 -m pip install nltk sacremoses pandas regex mock transformers>=4.33.2 mosestokenizer
!python3 -c "import nltk; nltk.download('punkt')"
!python3 -m pip install bitsandbytes scipy accelerate datasets
!python3 -m pip install sentencepiece

!git clone https://github.com/VarunGumma/IndicTransTokenizer
%cd IndicTransTokenizer
!python3 -m pip install --editable ./
%cd ..

In [ ]:
# !pip install IndicTransTokenizer

ERROR: Could not find a version that satisfies the requirement IndicTransTokenizer (from versions: none)
ERROR: No matching distribution found for IndicTransTokenizer


In [ ]:
import torch
from transformers import AutoModelForSeq2SeqLM, BitsAndBytesConfig
from IndicTransTokenizer import IndicProcessor, IndicTransTokenizer

BATCH_SIZE = 4
DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
quantization = None

In [ ]:
def initialize_model_and_tokenizer(ckpt_dir, direction, quantization):
    if quantization == "4-bit":
        qconfig = BitsAndBytesConfig(
            load_in_4bit=True,
            bnb_4bit_use_double_quant=True,
            bnb_4bit_compute_dtype=torch.bfloat16,
        )
    elif quantization == "8-bit":
        qconfig = BitsAndBytesConfig(
            load_in_8bit=True,
            bnb_8bit_use_double_quant=True,
            bnb_8bit_compute_dtype=torch.bfloat16,
        )
    else:
        qconfig = None

    tokenizer = IndicTransTokenizer(direction=direction)
    model = AutoModelForSeq2SeqLM.from_pretrained(
        ckpt_dir,
        trust_remote_code=True,
        low_cpu_mem_usage=True,
        quantization_config=qconfig,
    )

    if qconfig == None:
        model = model.to(DEVICE)
        if DEVICE == "cuda":
            model.half()

    model.eval()

    return tokenizer, model


def batch_translate(input_sentences, src_lang, tgt_lang, model, tokenizer, ip):
    translations = []
    for i in range(0, len(input_sentences), BATCH_SIZE):
        print(i)
        batch = input_sentences[i : i + BATCH_SIZE]

        # Preprocess the batch and extract entity mappings
        batch = ip.preprocess_batch(batch, src_lang=src_lang, tgt_lang=tgt_lang)

        # Tokenize the batch and generate input encodings
        inputs = tokenizer(
            batch,
            src=True,
            truncation=True,
            padding="longest",
            return_tensors="pt",
            return_attention_mask=True,
        ).to(DEVICE)

        # Generate translations using the model
        with torch.no_grad():
            generated_tokens = model.generate(
                **inputs,
                use_cache=True,
                min_length=0,
                max_length=256,
                num_beams=5,
                num_return_sequences=1,
            )

        # Decode the generated tokens into text
        generated_tokens = tokenizer.batch_decode(generated_tokens.detach().cpu().tolist(), src=False)

        # Postprocess the translations, including entity replacement
        translations += ip.postprocess_batch(generated_tokens, lang=tgt_lang)

        del inputs
        torch.cuda.empty_cache()

    return translations

In [ ]:
### English to Indic Example


en_indic_ckpt_dir = "ai4bharat/indictrans2-en-indic-1B"  # ai4bharat/indictrans2-en-indic-dist-200M
en_indic_tokenizer, en_indic_model = initialize_model_and_tokenizer(en_indic_ckpt_dir, "en-indic", quantization)

ip = IndicProcessor(inference=True)

# en_sents = [
#     "When I was young, I used to go to the park every day.",
#     "He has many old books, which he inherited from his ancestors.",
#     "I can't figure out how to solve my problem.",
#     "She is very hardworking and intelligent, which is why she got all the good marks.",
#     "We watched a new movie last week, which was very inspiring.",
#     "If you had met me at that time, we would have gone out to eat.",
#     "She went to the market with her sister to buy a new sari.",
#     "Raj told me that he is going to his grandmother's house next month.",
#     "All the kids were having fun at the party and were eating lots of sweets.",
#     "My friend has invited me to his birthday party, and I will give him a gift.",
# ]

src_lang, tgt_lang = "eng_Latn", "guj_Gujr"
print("startd")
eng_to_gu_mbart = batch_translate(english_sentences, src_lang, tgt_lang, en_indic_model, en_indic_tokenizer, ip)

# print(f"\n{src_lang} - {tgt_lang}")
# for input_sentence, translation in zip(en_sents, hi_translations):
#     print(f"{src_lang}: {input_sentence}")
#     print(f"{tgt_lang}: {translation}")

# flush the models to free the GPU memory
del en_indic_tokenizer, en_indic_model

config.json:   0%|          | 0.00/1.27k [00:00<?, ?B/s]

configuration_indictrans.py:   0%|          | 0.00/14.1k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/ai4bharat/indictrans2-en-indic-1B:
- configuration_indictrans.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


modeling_indictrans.py:   0%|          | 0.00/61.3k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/ai4bharat/indictrans2-en-indic-1B:
- modeling_indictrans.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


model.safetensors:   0%|          | 0.00/4.46G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/163 [00:00<?, ?B/s]

startd
0
4
8
12
16
20
24
28
32
36
40
44
48
52
56
60
64
68
72
76
80
84
88
92
96
100
104
108
112
116
120
124
128
132
136
140
144
148
152
156
160
164
168
172
176
180
184
188
192
196
200
204
208
212
216
220
224
228
232
236
240
244
248
252
256
260
264
268
272
276
280
284
288
292
296
300
304
308
312
316
320
324
328
332
336
340
344
348
352
356
360
364
368
372
376
380
384
388
392
396
400
404
408
412
416
420
424
428
432
436
440
444
448
452
456
460
464
468
472
476
480
484
488
492
496
500
504
508
512
516
520
524
528
532
536
540
544
548
552
556
560
564
568
572
576
580
584
588
592
596
600
604
608
612
616
620
624
628
632
636
640
644
648
652
656
660
664
668
672
676
680
684
688
692
696
700
704
708
712
716
720
724
728
732
736
740
744
748
752
756
760
764
768
772
776
780
784
788
792
796
800
804
808
812
816
820
824
828
832
836
840
844
848
852
856
860
864
868
872
876
880
884
888
892
896
900
904
908
912
916
920
924
928
932
936
940
944
948
952
956
960
964
968
972
980
984
988
992
996


In [ ]:
# Open a new text file for writing
with open('eng_to_gu_mbart.txt', 'w') as file:
    # Iterate over each string in the list
    for string in eng_to_gu_mbart:
        # Write the string to the file followed by a newline character
        file.write(string + '\n')

In [ ]:
### Indic to English Example

indic_en_ckpt_dir = "ai4bharat/indictrans2-indic-en-1B"  # ai4bharat/indictrans2-indic-en-dist-200M
indic_en_tokenizer, indic_en_model = initialize_model_and_tokenizer(indic_en_ckpt_dir, "indic-en", "")

ip = IndicProcessor(inference=True)

# hi_sents = [
#     "जब मैं छोटा था, मैं हर रोज़ पार्क जाता था।",
#     "उसके पास बहुत सारी पुरानी किताबें हैं, जिन्हें उसने अपने दादा-परदादा से विरासत में पाया।",
#     "मुझे समझ में नहीं आ रहा कि मैं अपनी समस्या का समाधान कैसे ढूंढूं।",
#     "वह बहुत मेहनती और समझदार है, इसलिए उसे सभी अच्छे मार्क्स मिले।",
#     "हमने पिछले सप्ताह एक नई फिल्म देखी जो कि बहुत प्रेरणादायक थी।",
#     "अगर तुम मुझे उस समय पास मिलते, तो हम बाहर खाना खाने चलते।",
#     "वह अपनी दीदी के साथ बाजार गयी थी ताकि वह नई साड़ी खरीद सके।",
#     "राज ने मुझसे कहा कि वह अगले महीने अपनी नानी के घर जा रहा है।",
#     "सभी बच्चे पार्टी में मज़ा कर रहे थे और खूब सारी मिठाइयाँ खा रहे थे।",
#     "मेरे मित्र ने मुझे उसके जन्मदिन की पार्टी में बुलाया है, और मैं उसे एक तोहफा दूंगा।",
# ]
src_lang, tgt_lang = "guj_Gujr", "eng_Latn"
gu_to_en_mbart = batch_translate(gujarati_sentences, src_lang, tgt_lang, indic_en_model, indic_en_tokenizer, ip)


# print(f"\n{src_lang} - {tgt_lang}")
# for input_sentence, translation in zip(hi_sents, en_translations):
#     print(f"{src_lang}: {input_sentence}")
#     print(f"{tgt_lang}: {translation}")

# flush the models to free the GPU memory
del indic_en_tokenizer, indic_en_model

config.json:   0%|          | 0.00/1.27k [00:00<?, ?B/s]

configuration_indictrans.py:   0%|          | 0.00/14.1k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/ai4bharat/indictrans2-indic-en-1B:
- configuration_indictrans.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


modeling_indictrans.py:   0%|          | 0.00/61.3k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/ai4bharat/indictrans2-indic-en-1B:
- modeling_indictrans.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


pytorch_model.bin:   0%|          | 0.00/4.09G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/163 [00:00<?, ?B/s]

0
4
8
12
16
20
24
28
32
36
40
44
48
52
56
60
64
68
72
76
80
84
88
92
96
100
104
108
112
116
120
124
128
132
136
140
144
148
152
156
160
164
168
172
176
180
184
188
192
196
200
204
208
212
216
220
224
228
232
236
240
244
248
252
256
260
264
268
272
276
280
284
288
292
296
300
304
308
312
316
320
324
328
332
336
340
344
348
352
356
360
364
368
372
376
380
384
388
392
396
400
404
408
412
416
420
424
428
432
436
440
444
448
452
456
460
464
468
472
476
480
484
488
492
496
500
504
508
512
516
520
524
528
532
536
540
544
548
552
556
560
564
568
572
576
580
584
588
592
596
600
604
608
612
616
620
624
628
632
636
640
644
648
652
656
660
664
668
672
676
680
684
688
692
696
700
704
708
712
716
720
724
728
732
736
740
744
748
752
756
760
764
768
772
776
780
784
788
792
796
800
804
808
812
816
820
824
828
832
836
840
844
848
852
856
860
864
868
872
876
880
884
888
892
896
900
904
908
912
916
920
924
928
932
936
940
944
948
952
956
960
964
968
972
976
980
984
988
992
996


In [ ]:
# Open a new text file for writing
with open('gu_to_en_mbart.txt', 'w') as file:
    # Iterate over each string in the list
    for string in gu_to_en_mbart:
        # Write the string to the file followed by a newline character
        file.write(string + '\n')

In [ ]:
### Indic to Indic Example


indic_indic_ckpt_dir = "ai4bharat/indictrans2-indic-indic-1B"  # ai4bharat/indictrans2-indic-indic-dist-320M
indic_indic_tokenizer, indic_indic_model = initialize_model_and_tokenizer(indic_indic_ckpt_dir, "indic-indic", quantization)

ip = IndicProcessor(inference=True)

# hi_sents = [
#     "जब मैं छोटा था, मैं हर रोज़ पार्क जाता था।",
#     "उसके पास बहुत सारी पुरानी किताबें हैं, जिन्हें उसने अपने दादा-परदादा से विरासत में पाया।",
#     "मुझे समझ में नहीं आ रहा कि मैं अपनी समस्या का समाधान कैसे ढूंढूं।",
#     "वह बहुत मेहनती और समझदार है, इसलिए उसे सभी अच्छे मार्क्स मिले।",
#     "हमने पिछले सप्ताह एक नई फिल्म देखी जो कि बहुत प्रेरणादायक थी।",
#     "अगर तुम मुझे उस समय पास मिलते, तो हम बाहर खाना खाने चलते।",
#     "वह अपनी दीदी के साथ बाजार गयी थी ताकि वह नई साड़ी खरीद सके।",
#     "राज ने मुझसे कहा कि वह अगले महीने अपनी नानी के घर जा रहा है।",
#     "सभी बच्चे पार्टी में मज़ा कर रहे थे और खूब सारी मिठाइयाँ खा रहे थे।",
#     "मेरे मित्र ने मुझे उसके जन्मदिन की पार्टी में बुलाया है, और मैं उसे एक तोहफा दूंगा।",
# ]
src_lang, tgt_lang = "guj_Gujr", "mar_Deva"
gu_to_mr_mbart = batch_translate(gujarati_sentences, src_lang, tgt_lang, indic_indic_model, indic_indic_tokenizer, ip)

# print(f"\n{src_lang} - {tgt_lang}")
# for input_sentence, translation in zip(hi_sents, mr_translations):
#     print(f"{src_lang}: {input_sentence}")
#     print(f"{tgt_lang}: {translation}")

# flush the models to free the GPU memory
del indic_indic_tokenizer, indic_indic_model



config.json:   0%|          | 0.00/1.28k [00:00<?, ?B/s]

configuration_indictrans.py:   0%|          | 0.00/14.1k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/ai4bharat/indictrans2-indic-indic-1B:
- configuration_indictrans.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


modeling_indictrans.py:   0%|          | 0.00/61.3k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/ai4bharat/indictrans2-indic-indic-1B:
- modeling_indictrans.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


pytorch_model.bin:   0%|          | 0.00/4.83G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/168 [00:00<?, ?B/s]

0
4
8
12
16
20
24
28
32
36
40
44
48
52
56
60
64
68
72
76
80
84
88
92
96
100
104
108
112
116
120
124
128
132
136
140
144
148
152
156
160
164
168
172
176
180
184
188
192
196
200
204
208
212
216
220
224
228
232
236
240
244
248
252
256
260
264
268
272
276
280
284
288
292
296
300
304
308
312
316
320
324
328
332
336
340
344
348
352
356
360
364
368
372
376
380
384
388
392
396
400
404
408
412
416
420
424
428
432
436
440
444
448
452
456
460
464
468
472
476
480
484
488
492
496
500
504
508
512
516
520
524
528
532
536
540
544
548
552
556
560
564
568
572
576
580
584
588
592
596
600
604
608
612
616
620
624
628
632
636
640
644
648
652
656
660
664
668
672
676
680
684
688
692
696
700
704
708
712
716
720
724
728
732
736
740
744
748
752
756
760
764
768
772
776
780
784
788
792
796
800
804
808
812
816
820
824
828
832
836
840
844
848
852
856
860
864
868
872
876
880
884
888
892
896
900
904
908
912
916
920
924
928
932
936
940
944
948
952
956
960
964
968
972
976
980
984
988
992
996


In [ ]:
# Open a new text file for writing
with open('gu_to_mr_mbart.txt', 'w') as file:
    # Iterate over each string in the list
    for string in gu_to_mr_mbart:
        # Write the string to the file followed by a newline character
        file.write(string + '\n')

In [ ]:
### Indic to Indic Example


indic_indic_ckpt_dir = "ai4bharat/indictrans2-indic-indic-1B"  # ai4bharat/indictrans2-indic-indic-dist-320M
indic_indic_tokenizer, indic_indic_model = initialize_model_and_tokenizer(indic_indic_ckpt_dir, "indic-indic", quantization)

ip = IndicProcessor(inference=True)

# hi_sents = [
#     "जब मैं छोटा था, मैं हर रोज़ पार्क जाता था।",
#     "उसके पास बहुत सारी पुरानी किताबें हैं, जिन्हें उसने अपने दादा-परदादा से विरासत में पाया।",
#     "मुझे समझ में नहीं आ रहा कि मैं अपनी समस्या का समाधान कैसे ढूंढूं।",
#     "वह बहुत मेहनती और समझदार है, इसलिए उसे सभी अच्छे मार्क्स मिले।",
#     "हमने पिछले सप्ताह एक नई फिल्म देखी जो कि बहुत प्रेरणादायक थी।",
#     "अगर तुम मुझे उस समय पास मिलते, तो हम बाहर खाना खाने चलते।",
#     "वह अपनी दीदी के साथ बाजार गयी थी ताकि वह नई साड़ी खरीद सके।",
#     "राज ने मुझसे कहा कि वह अगले महीने अपनी नानी के घर जा रहा है।",
#     "सभी बच्चे पार्टी में मज़ा कर रहे थे और खूब सारी मिठाइयाँ खा रहे थे।",
#     "मेरे मित्र ने मुझे उसके जन्मदिन की पार्टी में बुलाया है, और मैं उसे एक तोहफा दूंगा।",
# ]
src_lang, tgt_lang = "mar_Deva", "guj_Gujr"
mr_to_gu_mbart = batch_translate(marathi_sentences, src_lang, tgt_lang, indic_indic_model, indic_indic_tokenizer, ip)

# print(f"\n{src_lang} - {tgt_lang}")
# for input_sentence, translation in zip(hi_sents, mr_translations):
#     print(f"{src_lang}: {input_sentence}")
#     print(f"{tgt_lang}: {translation}")

# flush the models to free the GPU memory
del indic_indic_tokenizer, indic_indic_model



In [ ]:
# Open a new text file for writing
with open('mr_to_gu_mbart.txt', 'w') as file:
    # Iterate over each string in the list
    for string in mr_to_gu_mbart:
        # Write the string to the file followed by a newline character
        file.write(string + '\n')

# **1.3) Chat-GPT**

In [ ]:
# Read English sentences from test_en file
test_en_file = os.path.join("test.en")
english_sentences_gpt=[]
gujarati_sentences_gpt=[]
marathi_sentences_gpt=[]
with open(test_en_file, "r", encoding="utf-8") as file:
    english_sentences_gpt = [line.strip() for line, _ in zip(file.readlines(), range(50))]

test_gu_file = os.path.join("test.gu")
with open(test_gu_file, "r", encoding="utf-8") as file:
    gujarati_sentences_gpt = [line.strip() for line, _ in zip(file.readlines(), range(50))]

test_mr_file = os.path.join("test.mr")
with open(test_mr_file, "r", encoding="utf-8") as file:
    marathi_sentences_gpt = [line.strip() for line, _ in zip(file.readlines(), range(50))]

In [ ]:
eng_to_gu_chatGPT = [
    "આ ઘટનાના પરિપ્રેક્ષ્ય સાથે, હું આશા કરું છું કે વિદેશથી આવતા પ્રતિનિધિઓ કોઈ સમય દિલ્હીની ઈતિહાસ અને સ્પ્લેન્ડર જોવા માટે કેટલો સમય આપશે.",
    "આપણે 2018 ના વિશ્વ પર્યાવરણ દિવસના વૈશ્વિક આયોજક હોવાને ગર્વ છીએ.",
    "આપણે નિશ્ચિત કરવા માટે પ્રતિષ્ઠા છે કે અમે તે વૃદ્ધિ કરીએ છીએ, જે સાચું અને હરિત છે.",
    "આ પિરતી સ્મોકની દુઃખથી ગામડાના સ્ત્રીઓને મુકતિ મળી છે.",
    "આપણે નવી ઊર્જા ઉત્પાદનની દિશામાં મોટી મુદ્દત મુક્તાંગ છીએ.",
    "અમે વિશ્વની પાંચમી સૌર ઊર્જા ઉત્પાદક છીએ.",
    "ના કેવળ તે, પરંતુ અમે જિણેની છીએ નવી ઊર્જા ઉત્પાદક.",
    "અમે એક યુવા રાષ્ટ્ર છીએ.",
    "આપણી વૃક્ષ અને વનનો આચર્ય અંતર પ્રતિષ્ઠાનું વધાર્યું છે.",
    "અમે વન્ય જીવન સંરક્ષણમાં પણ ચડાવેલા છીએ.",
    "કૃષિ માટે પોતાના અનુસરણની ઉપલબ્ધતાનું ચાલુ રહેવા મહત્વપૂર્ણ છે.",
    "અમારું મોટ્ટો ઉદ્દીપન છે 'ઓછી જમીન પર વધુ વર્ષ, પ્રતિ ડ્રોપ.'",
    "વિશ્વ પર્યાવરણ દિવસ આ વર્ષ ખૂબ મહત્વની ચૂકવા માંગે છે.",
    "પ્લાસ્ટિક હવામાં હમણાં માનવતાની એક પ્રકાર બની રહ્યું છે.",
    "પ્લાસ્ટિક પ્રદૂષણ હવામાં હમણાં માનવના ખાદ્યપોષણમાં પ્રવેશ કરી રહ્યું છે.",
    "વાસ્તવમાં, માઇક્રો પ્લાસ્ટિક હવામાં પણ આવ્યું છે, જેમ કે ખાંડ, બોટલ પાણી અને હાથમાં પાણી.",
    "ભારતમાં પર કેપિટા પ્લાસ્ટિક વપરાશ કઈ કય કઈ જગ્યાઓ પર વધુ છે.",
    "હું આશા કરું છું કે તેઓ પ્લાસ્ટિક પ્રદૂષણને રોકવામાં ઉદાહરણીય કામ કરી રહ્યાં છે.",
    "જે પસંદીઓ આજ અમારો સામૂહિક ભવિષ્ય નિર્ધારશ કરશે.",
    "તે હાલ ભારત પર અંતરરાષ્ટ્રીય પ્રવાસ કરી રહ્યાં છે જે મુખ્ય પ્રયાસ 'વતન કો જાણો' ભાગીદારી છે.",
  "તેમની સંપર્ક કરવાથી યોજનાઓ વિશે ચર્ચા કરતે હોવા સાથે, પ્રધાનમંત્રી ને રાજ્યમાં સંપૃક્તતા અને ઇન્ફ્રાસ્ટ્રક્ચરને સુધારવા માટે કે યુનિયન સરકારના પ્રયાસો વિશે વાત કરી.",
    "આ વર્ષે કેલિફોર્નિયા, યુનાઇટેડ સ્ટેટ્સ ઓફ અમેરિકામાં સૌરાષ્ટ્ર પટેલ સાંસ્કૃતિક સમાજનો આંતરરાષ્ટ્રીય મેળો યોજાયો છે.",
    "પ્રધાનમંત્રી નેહરૂઓની પ્રયાસો પર થાપવામાં આવવી પછી, વિશેષ કરીને સૌરાષ્ટ્ર પટેલ સમુદાયની મહેનતને ભારત ગૌરવાની આભા પામી.",
    "તેમને ઉમેરે છે કે તેમની પ્રયાસોથી ભારતીય પાસપોર્ટ દુનિયાની દરેક જગ્યાએ માન્ય છે.",
    "પ્રધાનમંત્રીએ કેટલાક યોજનાઓ વિશે કહ્યું કે, જેમ કે સ્વચ્છ ભારત જેવી સરકારની વિવિધ યોજનાઓ દેશની પર્યટનને મોટી પ્રોત્સાહન આપ્યો છે.",
    "તે પણ બહારના વિદેશી કુટુંબોને ભારત દર્શન માટે પ્રેરિત કરવાની NRI સમુદાયને આહ્વાન આપ્યો.",
    "પ્રધાનમંત્રીએ કહ્યું કે ભારત ગાંધીજીની 150મી જન્મ જયંતીને ઓક્ટોબર 2માંથી જ ઉજવાયી છે.",
    "જીએસટી જેવા ઉદ્યોગો અને ભ્રષ્ટાચારવિરુદ્ધ કડક ક્રિયાઓની પણ મદદ કરવાથી ઈમાનદાર વ્યાપાર કરનારા લોકોને મદદ મળી છે.",
    "આ પ્રયત્નોએ વ્યાપાર કરવાની સુવિધામાં 4 વર્ષમાં 42 સ્થાન પર કૂદતી બિઝનેસ કરવાની માપદંડોમાં વૃદ્ધિ આવી છે, તે ઉમેર્યું.",
    "પ્રધાનમંત્રી આજ કચ્છમાં વિવિધ વિકાસ પ્રોજેક્ટ્સનું ઉદ્ઘાટન કરી નાંખવાનું છે અને આધારશિલા પર પથનિર્માણ કરી છે.",
    "તે આપણી પાસે તે દેશ માટે જીવવા માટે તેમની જીવન આપી છે.",
    "જેમણા પરિવારો ને તેમના નજીકના અને પ્રિય મિત્રો ને માર્યા છે, મનની શોકસંદેહનું સારું શોક્ષણું.",
    "આ ક્ષણે, પ્રતિક્રિયાની અપેક્ષાઓ અને સખત પ્રતિસાદની ભાવનાઓ છે જે બહુ પ્રાકૃતિક છે.",
    "અમે સિક્યુરિટી ફોર્સને પૂરી આજાદી આપી છીએ.",
    "અમારી સિક્યુરિટી ફોર્સના સાહસ અને વીરતાને અમે પૂરી શ્રદ્ધા ધરીએ છીએ.",
    "130 કરોડ ભારતીય જે પણ તેમના કેમ હોય એવા પ્રતિસાદને સંજોગી પાડવામાં આવશે.",
    "અમારા વીર સૈનિકો ને તેમની જીવન આપી છે.",
    "પ્રધાનમંત્રીએ પાકિસ્તાનના લોકોને કહ્યું કે તેમના શાસકોને ભ્રષ્ટાચાર અને નકલી નોટ્સ મુકાવવા માટે લડવાનું કહો.",
    "પ્રધાનમંત્રીએ કહ્યું કે યુનિયન સરકાર કૃષિપરિસંચાળન અને પૂરી શક્તિથી પાણી આપવા માટે કંઈક પણ કરી શકે છે.",
    "ભારતની પાસવર્ડ પાકિસ્તાન જવામાં ન જવું દેવામાં આવતું પાણી બંધ છે, તે ઉમેર્યું.",
    "પ્રધાનમંત્રીએ ટ્વીટ કર્યું કે કેન્દ્રે બિહાર સરકારને બાઢ સ્થિતિને શમાંત કરવા માટે બધી સહાય આપવામાં આવશે અને કે સ્થિતિને ઘણી નજીકે જુઓ રહ્યું છે.",
    "બિહારના ભાગોમાં બાઢના પ્રભાવિત થયેલ બંને લોકો સાથે મારી વિચારો છે.",
    "સ્થિતિને ઘણી નજીકે જુઓ રહ્યું છે.",
    "કેન્દ્ર બિહાર સરકારને બાઢ સ્થિતિને શમાંત કરવા માટે બધી સહાય આપશે એ તો સ્થિતિ બતાવી.",
    "પ્રધાનમંત્રીએ કાશીના લોકોને ફ્રેન્ચ પ્રમુખ એમાન્યુએલ મેક્રોનને આવક આપવા માટે આભાર આપ્યું.",
    "તે કહ્યું કે વારાણસીમાં પર્યટન માટે અત્યંત સંભાવના છે અને શહેરને સાફ રાખવા માટે બધા પ્રયત્નો કરવામાં આવી જોઈએ.",
    "પ્રધાનમંત્રીએ કહ્યું કે હાલમાં જાહેર કરવામાં આવેલ આયુષ્માન ભારત યોજના આરોગ્ય ખેતી વિપરીત કરવા માટે હળવાનું છે.",
    "તે ગરીબ લોકોને શ્રેષ્ઠ ગુણવત્તાવાળી આરોગ્ય સેવા પ્રદાન કરશે.",
    "શરૂઆતમાં, અસેહના હાલમાં જીવનના ક્ષતિને લેવાની હવે સાફ શોકાતો આપવા માટે અમી સંવેદના વ્યક્ત કરીશું.",
    "તમે એક મહાન રાષ્ટ્રના નેતા છો."
]


In [ ]:
gu_to_en_chatGPT = [
    "Along with this event, I hope that the representatives from abroad who are attending this gathering will also spend some time exploring the history and grandeur of Delhi.",
    "We have experienced the honor of being the global host for World Environment Day 2018.",
    "We are committed to raising the standard of living of our people.",
    "As a result, rural women have been liberated from difficult conditions of household smoke.",
    "We are accelerating renewable energy production significantly.",
    "We are already the fifth-largest producer of solar energy in the world.",
    "Furthermore, India is also the sixth-largest producer of renewable energy in the world.",
    "We are a youthful nation.",
    "In the last two years alone, we have seen a two-fold increase in our forest and tree cover.",
    "We are also doing commendable work in wildlife conservation.",
    "Therefore, continuous availability of water is crucial for agriculture.",
    "Our principle is 'increase productivity of every drop of water'.",
    "This year, a significant decision has been made to celebrate World Environment Day.",
    "Plastic has now become a severe pollution for human life.",
    "Plastic pollution now also enters our food chain.",
    "Microplastics are now found in sugary drinks, bottled water, and tap water.",
    "Compared to many developed countries, plastic consumption is much lower in India.",
    "I hope that efforts to control all this plastic pollution will set a good example.",
    "The choice we make today will determine our collective future.",
    "Currently, 'Know Your Country' is a major initiative traveling across various parts of India.",
    "During his interaction, the Prime Minister discussed efforts underway to improve connectivity and digital infrastructure in the state.",
    "The Saurashtra Patel Cultural Society's international conference is being held this year in California, USA.",
    "The Prime Minister praised the efforts of Indian diaspora, especially the Patel community of Saurashtra, stating that overseas Indians have always brought honor to India.",
    "He mentioned that the efforts of overseas Indians ensure respect for Indian passports worldwide.",
    "The Prime Minister stated that various initiatives of a clean India government have encouraged tourism in the country.",
    "He appealed to the overseas Indian community to inspire at least five foreign families to visit India every year.",
    "The Prime Minister announced that India will commence the celebration of Mahatma Gandhi's 150th birth anniversary from October 2.",
    "GST and initiatives against corruption have helped people conduct business with authenticity.",
    "He mentioned that as a result of these initiatives, India has jumped 42 places in ease of doing business rankings in the last four years.",
    "Today, the Prime Minister will inaugurate various development projects in Kutch and lay the foundation stone.",
    "He has sacrificed his own life in service to the nation.",
    "In this time of sorrow, my and every Indian's sentiments are with their families.",
    "Currently, there are expectations from the country, there is also a feeling of wanting to be relieved, which is natural as well.",
    "Our armed forces are given complete independence.",
    "We have complete confidence in the valor of our soldiers, their bravery.",
    "130 crore Indians will give a resounding response to every conspiracy of this kind, to every such attack.",
    "Our brave martyrs have sacrificed their lives.",
    "He appealed to the people of Pakistan to fight against corruption and counterfeit currency in front of their rulers.",
    "The Prime Minister said that the central government will be able to complete the irrigation work to provide adequate water to the farmers.",
    "India's water does not reach Pakistan, that's what he emphasized.",
    "The Prime Minister tweeted that the central government assures full cooperation to the Bihar state government to mitigate the situation and closely monitor the situation.",
    "The Prime Minister said, 'My thoughts are with the severely affected people in various parts of Bihar.",
    "The situation is continuously being monitored.",
    "The central government assures full cooperation to the Bihar government to mitigate the situation.'",
    "He praised the welcome by the residents of Kashi to the warm reception of French President Emmanuel Macron.",
    "He mentioned that there are promising possibilities for tourism in Varanasi and efforts should be made to keep the city clean.",
    "The Prime Minister further announced that the recently launched Ayushman Bharat Yojana is bringing changes in the healthcare sector.",
    "It will provide high-quality health services to the poor.",
    "In the beginning, I am saddened by the loss of life in recent earthquakes.",
    "You are a great leader of a great nation."
]

In [ ]:
gu_to_mr_chatGPT = [
    "त्यावरून मी आशा करतो की परदेशातील हा समावेश आणि संयुक्त प्रतिनिधी दिल्लीचे इतिहास आणि महत्त्व पाहण्यासाठीही थोडे वेळ घालायला लागेल.",
    "आम्ही जागतिक पर्यावरण दिवस 2018 दिवशाला जागतिक मेजरमध्ये भागविण्याची गरज अनुभवतो.",
    "आम्ही आमच्या लोकांच्या जीवनशैलीवर उंची घेण्यासाठी समर्पित आहोत.",
    "त्याचाच कारण, ग्रामीण स्त्रियोंना झेरी भंगणार्‍या स्थितीतून मुक्त झाली आहे.",
    "आम्ही पुन्हा ऊर्जा उत्पादनाला भारी वेगाने देत आहोत.",
    "आम्ही पहिल्यांदाच जगात सौर ऊर्जेच्या पाचव्या सर्वात मोठ्या उत्पादक देश आहोत.",
    "असं नाही, भारत अक्षय ऊर्जेच्या उत्पादनात जगातचा सहा था उत्पादक देश आहे.",
    "आम्ही युवा राष्ट्र आहोत.",
    "शेवटी, आपल्याला पडकाररूपात असेहीच प्रकार करण्यासाठी आज्ञा दिली जाते की वनस्पती व वनचरित्राच्या पाचव्या वाढ आली आहे.",
    "आम्ही वन्यजीवनाची वाट पडवत आहोत.",
    "त्याचाच कारण कृषीसाठी पाणीची सतत उपलब्धता महत्त्वपूर्ण आहे.",
    "आमचे सिद्धांत 'प्रत्येक बिंदूला जल देणारे वाढ' आहे.",
    "जागतिक पर्यावरण दिवस दिवशी ह्या वर्षी एक महत्वपूर्ण पडकार करण्याचा निर्णय घेण्यात आला आहे.",
    "प्लास्टिक आता मानव जीवनासाठी किंवा कचर्यात दुष्ट गोंधळ बनलेला आहे.",
    "प्लास्टिक प्रदूषण आता आपल्या अन्नामान्यामध्ये किंवा नलीच्या पाण्यामध्ये पण प्रवेश केलेला आहे.",
    "आता मायक्रो प्लास्टिक आपल्या मीठ्या, बोतलातील पाण्यामध्ये आणि नलीच्या पाण्यामध्ये पण प्रवेश केलेला आहे.",
    "अनेक विकसित देशांशी तुलना केल्यास भारतात प्लास्टिकचा वापर खूप कमी आहे.",
    "मी आशा करतो की हे सर्व प्लास्टिक प्रदूषण तसंच त्यांची नियंत्रण करण्यासाठी उदाहरण कार्यकीर्ती करेल.",
    "आपण आज ज्या प्रियकर्ता केल्या जातो, त्याचा समावेश आपल्या सर्वांच्या सामाजिक भावना नक्की करेल.",
    "आपण आता भारताच्या विविध भागात भ्रमण करण्याचा निर्णय केला आहे.",
    "श्रीमंत्रीने राज्यात कनेक्टिव्हिटी आणि मालखागत सुविधांतील सुधार करण्यासाठी जे प्रयत्न केले आहेत त्यांच्या सहाय्यासाठी सरकारचा पूर्ण सहकार्य अनुशंसित केला.",
"सौराष्ट्र पटेल सांस्कृतिक समाजाचं आंतरराष्ट्रीय सम्मेलन वर्षाला अमेरिकेतील कॅलिफोर्नियामध्ये आयोजित केलं आहे.",
"प्रधानमंत्रीने प्रवासी भारतीयांना, विशेषतः सौराष्ट्र पटेल समुदायांच्या प्रयत्नांची प्रशंसा केली आणि म्हणाले की, प्रवासी भारतीय हमेशा भारताला गौरवाने लाभलं आहे.",
"ते म्हणाले की, प्रवासी भारतीयांच्या प्रयत्नांमुळे जगातील सर्व ठिकाणी भारतीय पासपोर्टला सम्मान निश्चित झालं आहे.",
"प्रधानमंत्रीने म्हणाले की, स्वच्छ भारत यासारख्या सरकारच्या विविध योजनांच्या माध्यमातून प्रयात्न क्षेत्राला प्रोत्साहन मिळाले आहे.",
"ते प्रवासी भारतीय समुदायाला ओळखले की, प्रत्येक वर्षी ओलांडलेल्या पंच विदेशी कुटुंबांनी भारतदर्शनासाठी प्रेरित करावे.",
"प्रधानमंत्रीने म्हणाले की, भारत 2 ऑक्टोबरपासून महात्मा गांधींचा 150वा जन्मजयंतीचा उत्सव सुरू होईल.",
"जीएसटीसारख्या कार्यक्रमांनी आणि भ्रष्टाचाराच्या मामल्यांमध्ये कठोर प्रक्रियेतून लोकांना विश्वास संपूर्ण व्यवसाय करण्यास मदत मिळाली आहे.",
"ते म्हणाले की, या कार्यक्रमांच्या परिणामाने शेवटी 4 वर्षांमध्ये व्यापार-वाणिज्यात सोप्पीच्या क्रमांकात भारतने 42 स्थानांवर हानी भरली आहे.",
"प्रधानमंत्री आज कच्छमध्ये विविध विकासलक्षी प्रोजेक्ट्सचा उद्घाटन करेल आणि शिलारोपण करेल.",
"ते देशसेवा करून पोताने प्राणांचा बलिदान केला आहे.",
"दुःखातील आज वेळा माझ्या आणि प्रत्येक भारतीयाच्या संवेदना त्यांच्या कुटुंबांसह.",
"अत्यंतपणे देशातल्या प्रत्येकांची अपेक्षा, काही करी छूट देण्याची भावना, ती पण स्वाभाविक आहे.",
"आम्हाला आमच्या सैन्यदलांना पूर्ण स्वतंत्रता देण्यात आलेली आहे.",
"आम्हाला आमच्या सैन्यजवळच्या शौर्यावर, त्यांच्या वीरत्वावर पूर्ण विश्वास आहे.",
"१३० कोटी हिंदूंच्या ही प्रकारनी दर षडयंत्र, दर उमलेलं हलकं उत्तर देईल.",
"आमच्या वीर शहीदांनी स्वतःच्या प्राणांचा बलिदान दिला आहे.",
"ते पाकिस्तानच्या लोकांना त्यांच्या शासकांना भ्रष्टाचार आणि नकली नोटांबाबत लढवायची अपील केली होती.",
"प्रधानमंत्रीने म्हणाले की, केंद्र सरकार खर्चाचे पूर्ण प्राणी लावून पाणी पूर्ण करण्यास सर्व कामगिरी करेल.",
"भारताचं पाणी पाकिस्तानला मिळत नाही, त्याचं ते खात्री केलं होतं.",
    "प्रधानमंत्रींनी ट्विट केलं होतं की, केंद्र सरकारने बिहार सरकारला विकट स्थितीला कमी करण्यात मदत देण्याची खात्री दिली आहे आणि स्थितीवर काळसर नेण्यात आली आहे.",
    "प्रधानमंत्रींनी सांगितलं की, 'माझे विचार बिहारच्या विविध प्रदेशांमधील पूर्णपणे प्रभावित लोकांसोबत आहेत.",
    "स्थितीवर काळसर नेण्यात आली आहे.",
    "केंद्र सरकारने बिहार सरकारला संभाव्य सर्व सहाय्याची खात्री दिली आहे.",
    "त्यांनी फ्रान्सच्या राष्ट्रपती इमान्युअल माक्रोनला काशीचं उष्माभर्या स्वागत केलं.",
    "त्यांनी सांगितलं की, वाराणसीमध्ये प्रवासासाठी पुष्कळ संभावना आहेत आणि शहराची स्वच्छता ठेवण्यासाठी सर्व प्रयत्न करण्याची आवश्यकता आहे.",
    "प्रधानमंत्रींनी अधिक सांगितलं की, ताजे घोषणा केलेल्या आयुष्यमान भारत योजनेमध्ये आरोग्य क्षेत्रात परिवर्तन आणणार आहे.",
    "ते गरीबांना उच्च गुणवत्ताची आरोग्य सेवा प्रदान करेल.",
    "प्रारंभात मी आकेलेही ताजे धरतीकंपात झालेल्या जनामध्ये दुःखाची व्यक्ती करण्याची इच्छा असते.",
    "तुम्ही एक महान राष्ट्राचे नेते आहात."
]

In [ ]:
mr_to_gu_chatGPT = [
    "આ વિશેષ કાર્યક્રમ માટે પરદેશમાંથી આવતા પ્રતિનિધિઓને મારી વિનંતી છે કે, તેઓ આપલા આવા-યામમાં દિલ્લીનું વૈભવશીલ ઇતિહાસ ધરાવતી વાસ્તુઓને ભેટ આપવા માટે થોડા સમય જ ધરાવો.",
    "'વિશ્વ પર્યાવરણ દિન, 2018' આયોજન કરવાનો યજમાનપદ મળ્યો છે અમને માણસાતી લાગે છે.",
    "અર્થાત, શાશ્વત અને હરિત પ્રક્રિયાને વિકાસની દ્રષ્ટિને લાગુ કરવા માટે ભારત જોડાયેલું છે.",
    "તેથી ગામનાં મહિલાઓની સ્વતંત્રતા મળી ગઈ છે.",
    "નવીકરણશીલ ઊર્જા ઉત્પાદન માટેની મુખ્ય સંકલ્પની આમે કરી છીએ.",
    "સૌર ઊર્જા ખેત્રમાં વિશ્વમાં સૌથી વધુ વીજની ઉત્પાદક દેશોમાં અમે પાંચમાં સ્થાને છીએ.",
    "તેવી સમજ નથી તો અમે જાગતિક નવીકરણશીલ ઊર્જા ઉત્પાદન કરતી વેલી વિશ્વપટળે મોટું દેશ છીએ.",
    "આમારો દેશ તરુણ છે.",
    "ગયા વર્ષોમાં ભારતના વનમાંથી ભૂભાગ એક ટક્કા વધી ગયો છે.",
    "વન્યજીવ સંરક્ષણ પ્રદેશમાં પણ ભારતને ખૂબ ચાલુ કામ મળ્યું છે.",
    "નિરંતર જળ ઉપલબ્ધતા ખેતી માટે મહત્વનું છે અને તે માટે પ્રધાનમંત્રી કૃષિ સિંચાઈ યોજના શરૂ થઈ છે.",
    "આમાં આપણે 'મોર ક્રોપ, પર ડ્રોપ' જેવું ધ્યેય રાખીએ.",
    "'જાહેરાત દિવસ' ની યોજના મૂલ્યાંકન પ્રગટાવવાનો અધિકાર અમારી માટે હાજરીશીલ છે, આ ઓળખણી આપણે પ્રાપ્ત કરીશું અને તે અનુયાયી વર્ષે કરવાનું આવશ્યક છે.",
    "માનવને નિર્માણ કરીને પ્લાસ્ટિક દ્વારા પ્રદૂષણનું ધોકો થતો તેમ અધિક સચેત થવું જરૂરી છે.",
    "પ્લાસ્ટિક પ્રદૂષણ અને આપણની ખોરાકનો એક ભાગ થયો છે.",
    "મીઠુ, બોટલબંદ પાણી અને નળીની પાણી માટેની અતિ સૂક્ષ્મ ઘટકોથી બનેલ પ્લાસ્ટિકનો ઉપયોગ થતો હતો.",
    "વિકસિત જગ્યાઓની તુલનામાં ભારતમાં પ્લાસ્ટિકનો ઉપયોગ ખૂબ ઓછો છે.",
    "પ્લાસ્ટિક પ્રદૂષણ નિયંત્રણ માટે આ રીતે આદર્શ પ્રયાસ કરવામાં આવશે, આ આશા હું વ્યક્ત કરું છું.",
    "હવે તમે શું સારું કરવાની યોજના છે, તેનું સંયુક્ત પરિણામ તમારી ભવિષ્યને આનંદિત કરશે.",
    "'વતન નો જાનો' ની તળમજૂરીને આ મુલ્યાંકને આપણે તમારા દેશના વિવિધ ભાગોને મીતો છે.",
    "આ મુલાંનો સંવાદ કરતાં પ્રધાનમંત્રીએ જમ્મુ-કાશ્મીરમાં મુખ્ય પ્રવિધિઓ અને જોડણી માટે કેન્દ્ર સરકારની દ્વારા કરવામાં આવેલ પ્રકારના યોજનાઓની માહિતી આપી.",
    "અમેરિકામાં કેલિફોર્નિયામાં આંતરરાષ્ટ્રીય પરિષદ આવેલ છે.",
    "વિદેશવાસી ભારતીય વિશેષતઃ સૌરાષ્ટ્ર પટેલ સમાજના પ્રયાસોની પ્રશંસા કરતાં પ્રધાનમંત્રીએ કહ્યું કે, વિદેશવાસી ભારતીયોએ હંમેશા ભારતને અભિમાનથી વાતેલું એવું કામ કર્યું છે.",
    "તેમણે તેમના પ્રયાસોથી ભારતીય પાસપોર્ટને પ્રત્યેક સ્થળે આદર કર્યું છે, એવું કહ્યું.",
    "સ્વચ્છ ભારત યોજનાઓથી દેશના પર્યટનને ચાલુ થયું છે, એવું પ્રધાનમંત્રીએ કહ્યું.",
    "તેઓ વર્ષે વિદેશવાસી ભારતીય સમુદાયને પ્રતિ વર્ષે ઓછામાં ઓછી પાંચ પરદેશી કુટુંબોને ભારતમાં મેળવવા માટે પ્રોત્સાહિત કરવાનો આવહાન કર્યો.",
    "2 ઑક્ટોબરથી મહાત્મા ગાંધીની 150મી જયંતી મનાવવામાં આવશે, એવું પ્રધાનમંત્રીએ કહ્યું.",
    "જીએસટી જેવા પ્રકારના ઉપક્રમ અને ભ્રષ્ટાચારવિરુદ્ધ કઠોર પગલાં થવાથી લોકોને પ્રામાણિકપણે વ્યવસાય કરવા માટે પ્રેરિત કરવા માટે જરૂરી બનાવેલ છે.",
    "આ ઉપક્રમો દ્વારા છે વર્ષોમાં વ્યવસાય સુલભતાની માન્યતામાં ભારતે 42 સ્થળોને જીતી લીધું છે, એવું કહ્યું.",
    "કચ્છમાં વિવિધ વિકાસ કાર્યોની પાયાભૂતી, પ્રધાનમંત્રીએ કહ્યું કે હવે થશે.",
    "તેઓ દેશની સેવા કરતાં આપણા પ્રાણો ગુમાવ્યા છે.",
    "આ દુઃખના સંદર્ભમાં મારી અને દરેક ભારતીયની ભાવના તેમના પરિવારના સાથે છે.",
    "આ સમયે દેશની શું આશા છે, કંઈ કરીને બતાવવાની ભાવના, તે સ્વાભાવિક છે.",
    "અમારા સેનામાંના જવાનોને પૂર્ણ સ્વાતંત્ર્ય મેળવી ગયું હતું.",
    "સેનાની વીરતા, બહાદુરી પર મારો પૂર્ણ વિશ્વાસ છે.",
    "130 કરોડ ભારતીયો એવી દ્રાક્ષિકોને જવાબ આપીશું.",
    "અમારા વીર શહીદોએ પ્રાણોની બલિ આપી છે.",
    "તેઓ પાકિસ્તાનના લોકોને ભ્રષ્ટાચાર અને બનાવટ નોટાંવિરુદ્ધ તેમના રાજ્યને લડવાનો આહ્વાન આપ્યો છે.",
    "શેતકરોને પુરેસે પાણી મોકલવા માટે કેન્દ્ર સરકારે સાર્વત્રિક પ્રયત્ન કરે છે.",
    "ભારતની પાણી પાકિસ્તાનમાં જવા નહીં અસે, એવું તેઓ કહ્યું છે.",
    "પ્રધાનમંત્રીને ટ્વીટર સંદેશમાં મનાવવું છે કે, સ્થિતિને નિયંત્રિત કરવા માટે કેન્દ્ર સરકારે બિહાર સરકારને ઉચ્ચતમ મદદની આશ્વાસન આપ્યું છે અને સ્થિતિ પર ધ્યાન આપવો પડે છે.",
    "પ્રધાનમંત્રી માન્યવરે, 'બિહારના પૂરગ્રસ્તાંની દુઃખભરી સમસ્યાઓમાં હું ભાગીદાર છું.",
    "સ્થિતિ પર ધ્યાન આપવો પડે છે.",
    "સ્થિતિને નિયંત્રિત કરવા માટે કેન્દ્ર સરકારે બિહાર સરકારને ઉચ્ચતમ મદદની આશ્વાસન આપ્યું છે.",
    "ફ્રાન્સના રાષ્ટ્રાધ્યક્ષ એમાન્યુએલ મૅક્રોનનું હૃદયપૂર્વક સ્વાગત કરવામાં આવ્યું છે પ્રધાનમંત્રી મોદીએ કાશીના લોકોની આભાર માન્યું.",
    "વારાણસીમાં પર્યટન માટે મોટું શહેર ચાલવા માટે અને શહેરને સ્વચ્છ રાખવા માટે ઉચ્ચતમ પ્રયાસ કરવામાં આવ્યું છે અને એ પ્રધાનમંત્રીએ કહ્યું.",
    "'આયુષ્માન ભારત' યોજના હાલમાં શરૂ થઈ ગઈ છે આરોગ્ય ક્ષેત્રમાં ફેરફાર લાવવા માટે.",
    "આ યોજનાથી ગરીબોને ઉત્તમ આરોગ્ય સુવિધાઓ મળશે અને તે પ્રધાનમંત્રીએ કહ્યું.",
    "'અસે' પેટિકામાં ઘટણાસ્થળે બુધવારે ભૂકંપમાં જીવંત ખોટ થયેલો છે તે મને મનાપાસૂન શ્રદ્ધાંજલિ આપવો.",
    "તમે એક શ્રેષ્ઠ રાષ્ટ્રના નેતા છો."
]

# **Computing BLUE and ROUGE scores for NLLB-200 with 600M parameters:-**

### **1) English to Gujarati Translation**

In [ ]:
compute_BLEU_score(eng_to_gu,gujarati_sentences)
compute_ROUGE_score(gujarati_sentences,eng_to_gu)

BLEU Score: 0.5210612424722774
ROUGE Scores: {'rouge-1': {'r': 0.4200043965318141, 'p': 0.4803164436584559, 'f': 0.44199881830267707}, 'rouge-2': {'r': 0.1951016415078058, 'p': 0.2275819867736026, 'f': 0.20684755802184288}, 'rouge-l': {'r': 0.3977979517645144, 'p': 0.454876667965863, 'f': 0.418568371292865}}


### **2)Gujarati to English Translation:-**

In [ ]:
compute_BLEU_score(gu_to_en,english_sentences)
compute_ROUGE_score(english_sentences,gu_to_en)

BLEU Score: 0.658332028897129
ROUGE Scores: {'rouge-1': {'r': 0.5811043238802458, 'p': 0.5963906313030085, 'f': 0.5835050849999867}, 'rouge-2': {'r': 0.3515448564341267, 'p': 0.35959636384827093, 'f': 0.3519427338179226}, 'rouge-l': {'r': 0.5483304867626287, 'p': 0.562542502226389, 'f': 0.5505389370959229}}


### **3)Gujarati to Marathi Translation:-**

In [ ]:
compute_BLEU_score(gu_to_mr,marathi_sentences)
compute_ROUGE_score(marathi_sentences,gu_to_mr)

BLEU Score: 0.5215110581156264
ROUGE Scores: {'rouge-1': {'r': 0.36271924309013337, 'p': 0.3932708165974029, 'f': 0.37268280444687907}, 'rouge-2': {'r': 0.15470261612370603, 'p': 0.1664091378221025, 'f': 0.15824955067683313}, 'rouge-l': {'r': 0.33505325293490235, 'p': 0.3624657187755488, 'f': 0.34385685000057353}}


### **4)Marathi to Gujarati Translation:-**

In [ ]:
compute_BLEU_score(mr_to_gu,gujarati_sentences)
compute_ROUGE_score(gujarati_sentences,mr_to_gu)

BLEU Score: 0.008556708027480486
ROUGE Scores: {'rouge-1': {'r': 0.005295955704449948, 'p': 0.006645843922159709, 'f': 0.005806213910904373}, 'rouge-2': {'r': 0.00013333333333333334, 'p': 0.00018181818181818183, 'f': 0.0001538461489644972}, 'rouge-l': {'r': 0.005175955704449948, 'p': 0.006445843922159708, 'f': 0.005656213910904374}}


# **Computing BLUE and ROUGE scores for  IndicTrans:-**

### **1) English to Gujarati Translation**

In [ ]:
compute_BLEU_score(eng_to_gu_mbart,gujarati_sentences)
compute_ROUGE_score(gujarati_sentences,eng_to_gu_mbart)

BLEU Score: 0.6395359293665239
ROUGE Scores: {'rouge-1': {'r': 0.5005493957867624, 'p': 0.5111111995233298, 'f': 0.5025085918325749}, 'rouge-2': {'r': 0.25497334433471314, 'p': 0.2597864775090596, 'f': 0.25561221550922597}, 'rouge-l': {'r': 0.4726650854910783, 'p': 0.48235093940285123, 'f': 0.4743775789078798}}


### **2)Gujarati to English Translation:-**

In [ ]:
compute_BLEU_score(gu_to_en_mbart,english_sentences)
compute_ROUGE_score(english_sentences,gu_to_en_mbart)

BLEU Score: 0.728290636428684
ROUGE Scores: {'rouge-1': {'r': 0.6416228174604995, 'p': 0.6311766027880116, 'f': 0.6324123061900071}, 'rouge-2': {'r': 0.4245077158149204, 'p': 0.414477251160178, 'f': 0.41658616468957044}, 'rouge-l': {'r': 0.6091392214191292, 'p': 0.5996403259312879, 'f': 0.6007546157384641}}


### **3)Gujarati to Marathi Translation:-**

In [ ]:
compute_BLEU_score(gu_to_mr_mbart,marathi_sentences)
compute_ROUGE_score(marathi_sentences,gu_to_mr_mbart)

BLEU Score: 0.5881691988711897
ROUGE Scores: {'rouge-1': {'r': 0.4028044092452598, 'p': 0.3971730226363637, 'f': 0.395940276410292}, 'rouge-2': {'r': 0.17852849784185387, 'p': 0.17305867850987364, 'f': 0.17401245478743785}, 'rouge-l': {'r': 0.37301384917119496, 'p': 0.3678173837160582, 'f': 0.36669010562616666}}


### **4)Marathi to Gujarati Translation:-**

In [ ]:
# mr_to_gu_mbart=[]
# with open('mr_to_gu_mbart.txt', "r", encoding="utf-8") as file:
#     mr_to_gu_mbart = [line.strip() for line in file.readlines()]

In [ ]:
compute_BLEU_score(mr_to_gu_mbart,gujarati_sentences)
compute_ROUGE_score(gujarati_sentences,mr_to_gu_mbart)

BLEU Score: 0.5763753620053822
ROUGE Scores: {'rouge-1': {'r': 0.4278586744946865, 'p': 0.4382085893901877, 'f': 0.42826238840498404}, 'rouge-2': {'r': 0.19556714654763246, 'p': 0.19963120954220012, 'f': 0.19509255945395873}, 'rouge-l': {'r': 0.4031712079660919, 'p': 0.4126576046359492, 'f': 0.40346967242141646}}


# **Computing BLUE and ROUGE scores for  ChatGPT:-**

### **1) English to Gujarati Translation**

In [ ]:
compute_BLEU_score(eng_to_gu_chatGPT,gujarati_sentences_gpt)
compute_ROUGE_score(gujarati_sentences_gpt,eng_to_gu_chatGPT)

BLEU Score: 0.4675211370325956
ROUGE Scores: {'rouge-1': {'r': 0.28822533168585796, 'p': 0.295050219351101, 'f': 0.28830826707008295}, 'rouge-2': {'r': 0.09332949082296235, 'p': 0.09019499474591008, 'f': 0.09088495866078873}, 'rouge-l': {'r': 0.27095080066132693, 'p': 0.2779690505199321, 'f': 0.27132046710559643}}


### **2)Gujarati to English Translation:-**

In [ ]:
len(gu_to_en_chatGPT)

50

In [ ]:
compute_BLEU_score(gu_to_en_chatGPT,english_sentences_gpt)
compute_ROUGE_score(english_sentences_gpt,gu_to_en_chatGPT)

BLEU Score: 0.6109299502229472
ROUGE Scores: {'rouge-1': {'r': 0.5216103142493977, 'p': 0.5414481877285754, 'f': 0.5276434377434175}, 'rouge-2': {'r': 0.27381572455453196, 'p': 0.28368744443488686, 'f': 0.27633614917356886}, 'rouge-l': {'r': 0.47529795272031783, 'p': 0.49225072443298323, 'f': 0.48029362324227504}}


### **3)Gujarati to Marathi Translation:-**

In [ ]:
len(gu_to_mr_chatGPT)

50

In [ ]:
compute_BLEU_score(gu_to_mr_chatGPT,marathi_sentences_gpt)
compute_ROUGE_score(marathi_sentences_gpt,gu_to_mr_chatGPT)

BLEU Score: 0.45179687295095133
ROUGE Scores: {'rouge-1': {'r': 0.23488456001903302, 'p': 0.23310053908640532, 'f': 0.23166260098544114}, 'rouge-2': {'r': 0.06487909312909312, 'p': 0.0612962536782701, 'f': 0.06253076512551116}, 'rouge-l': {'r': 0.2039527532636968, 'p': 0.2021225375557722, 'f': 0.20085980000138967}}


### **4)Marathi to Gujarati Translation:-**

In [ ]:
len(mr_to_gu_chatGPT)

50

In [ ]:
compute_BLEU_score(mr_to_gu_chatGPT,gujarati_sentences_gpt)
compute_ROUGE_score(gujarati_sentences_gpt,mr_to_gu_chatGPT)

BLEU Score: 0.45543552320190417
ROUGE Scores: {'rouge-1': {'r': 0.2566881568723674, 'p': 0.2790976622411405, 'f': 0.26376900656822483}, 'rouge-2': {'r': 0.06863156399843706, 'p': 0.07891717171717172, 'f': 0.07279578308398915}, 'rouge-l': {'r': 0.22278631894421366, 'p': 0.2412831989266772, 'f': 0.22870745539308238}}


# **Question 3**

# **Key Learnings:**
## **Model Evaluation and Performance:**

### **NLLB-200 (600M parameters):**

- Performs moderately well for English to Gujarati translation (BLEU: 0.52, ROUGE: moderate).
- Achieves better results for Gujarati to English translation (BLEU: 0.66, ROUGE: good).
- Struggles significantly for Marathi to Gujarati translation (low BLEU and ROUGE scores).

### **IndicTrans:**

- Demonstrates good performance across different translation tasks, especially for Gujarati to English (high BLEU and ROUGE scores).
- Shows promising results for English to Gujarati translation as well.
- Performs decently for Gujarati to Marathi and Marathi to Gujarati translations.

### **ChatGPT:**

- Provides reasonable translation quality, with varying performance across different translation tasks.
- Scores comparatively lower BLEU and ROUGE scores compared to IndicTrans in most cases.

## **Understanding BLEU and ROUGE Scores:**

### **BLEU (Bilingual Evaluation Understudy):**

- Measures the similarity between the generated translation and one or more reference translations based on n-gram overlap.
- Higher BLEU scores indicate better translation quality, focusing on the precision of n-gram matches.
- BLEU score emphasizes the literal word-for-word overlap between the generated text and the reference text.
- Limitations include not capturing semantic similarity and being sensitive to word order.


### **ROUGE (Recall-Oriented Understudy for Gisting Evaluation):**
- It is a set of metrics used for evaluating the quality of text summaries or translations by comparing them to reference texts. It focuses on measuring the overlap of n-gram sequences and word sequences between the generated text and the reference text.

####**ROUGE scores include:**

- **ROUGE-1**: Measures unigram overlap, capturing how many unigrams (individual words) in the generated text match those in the reference text.
- **ROUGE-2**: Measures bigram overlap, assessing how many bigram sequences (pairs of adjacent words) are shared between the generated text and the reference text.
- **ROUGE-L**: Computes the longest common subsequence (LCS) of words between the generated text and the reference text, emphasizing longer word sequences.

####**Each ROUGE score consists of precision (p), recall (r), and F1-score (f):**

- **Precision (p)**: Indicates the proportion of generated n-grams that appear in the reference text.
- **Recall (r)**: Reflects the proportion of reference n-grams that are present in the generated text.
- **F1-score (f)**: Harmonic mean of precision and recall, providing a balanced measure of how well the generated text matches the reference text.

### **BLEU Score:**
#### **NLLB-200:**
- English to Gujarati (BLEU: 0.52): Moderate quality translation with decent overlap.
- Gujarati to English (BLEU: 0.66): Good quality translation with substantial overlap.
- Gujarati to Marathi (BLEU: 0.52): Moderate translation quality.
- Marathi to Gujarati (BLEU: 0.01): Very low quality translation.

#### **IndicTrans:**
- English to Gujarati (BLEU: 0.64): Good translation quality with significant overlap.
- Gujarati to English (BLEU: 0.73): High translation quality with substantial overlap.
- Gujarati to Marathi (BLEU: 0.59): Good translation quality.
- Marathi to Gujarati (BLEU: 0.58): Decent translation quality.

#### **ChatGPT:**
- English to Gujarati (BLEU: 0.47): Moderate translation quality.
- Gujarati to English (BLEU: 0.61): Good translation quality.
- Gujarati to Marathi (BLEU: 0.45): Moderate translation quality.
- Marathi to Gujarati (BLEU: 0.46): Moderate translation quality.

### **ROUGE Score:**
#### **Insights:**
- Generally, higher ROUGE scores (closer to 1) indicate better overlap and similarity with reference translations.

#### **Model Comparison:**
- IndicTrans generally performs well across different translation tasks compared to NLLB-200 and ChatGPT.
- NLLB-200 shows variability in performance depending on the translation direction.
- ChatGPT performs moderately but consistently across translation tasks.

#### **Translation Directions:**
- English to Gujarati and Gujarati to English translations tend to have higher BLEU and ROUGE scores, indicating relatively better quality.
- Translations involving Marathi show lower scores, suggesting challenges in translating between certain Indian languages.

## **Model Performance Insights:**
- The choice of model significantly impacts translation quality across different language pairs.
- The scores suggest that IndicTrans performs competitively in translating between English and Indian languages.
- Translation quality varies significantly based on the translation direction and the specific language pairs involved.
- BLEU and ROUGE scores provide valuable insights into the strengths and weaknesses of machine translation models for Indian languages.


In summary, this assignment taught us a lot about how well different machine translation models work for Indian languages. We learned how important it is to use metrics like BLEU and ROUGE to judge the quality of translations and choose the right model for different tasks. Continued research and improvement in this area are crucial for making machine translation even better for multiple languages.
